### PR TITLE
nav_transition_automation

### PR DESCRIPTION
**Pull request by dewagter**
Final PR for quadplane rotwing

### CODE REVIEW

File: sw/airborne/modules/rot_wing_drone/rot_wing_automation.h [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+ * @author Dennis van Wijngaarden <D.C.vanWijngaarden@tudelft.nl>
+ * Fucntions to automate the navigation and guidance of the rotating wing drone
+ */
+
+#ifndef ROT_WING_AUTOMATION_H
+#define ROT_WING_AUTOMATION_H
+
+#include "std.h"
+#include "math/pprz_algebra_float.h"
+#include "modules/rot_wing_drone/rotwing_state.h"
```
Comment (gautierhattenberger) [inline sw/airborne/modules/rot_wing_drone/rot_wing_automation.h:31]: this file doesn't exist

File: conf/airframes/tudelft/rot_wing_25kg.xml [hunk 2]
```diff
@@ -393,6 +422,11 @@
     </section>
 
     <section name="GUIDANCE_INDI_HYBRID" prefix="GUIDANCE_INDI_">
+        <!--WLS settings-->
+        <define name="USE_WLS"              value="FALSE|TRUE"/>
```
Comment (gautierhattenberger) [inline conf/airframes/tudelft/rot_wing_25kg.xml:426]: you let the compiler choose true or false ? ;)

**Review by gautierhattenberger — CHANGES_REQUESTED** [review]

File: sw/airborne/modules/rot_wing_drone/rot_wing_automation.h (hunk 1 above)
Comment (dewagter) [reply sw/airborne/modules/rot_wing_drone/rot_wing_automation.h:31]: fixed

File: conf/airframes/tudelft/rot_wing_25kg.xml (hunk 2 above)
Comment (dewagter) [inline conf/airframes/tudelft/rot_wing_25kg.xml:426]:
```suggestion
        <define name="USE_WLS"              value="TRUE"/>
```

Comment (dewagter) [reply conf/airframes/tudelft/rot_wing_25kg.xml]: Funny indeed, but it worked.

File: conf/airframes/tudelft/rot_wing_25kg.xml [hunk 3]
```diff
@@ -101,15 +106,25 @@
...
         <module name="guidance" type="indi_hybrid_quadplane"/>
-        <module name="nav_hybrid"/>
+        <module name="nav" type="hybrid">
+          <define name="GUIDANCE_H_USE_REF" value="FALSE"/>
+        </module>
 
         <!-- Other -->
         <module name="sys_id_doublet"/>
         <module name="sys_id_auto_doublets"/>
+        <module name="wing_rotation_controller_can"/>
```
Comment (dewagter) [inline conf/airframes/tudelft/rot_wing_25kg.xml:121]:
```suggestion
```

File: conf/airframes/tudelft/rot_wing_25kg.xml [hunk 4]
```diff
@@ -402,13 +435,22 @@
...
 
         <define name="THRUST_Z_EFF"         value="-0.0023"/>
         <define name="THRUST_X_EFF"         value="0.00055"/>
 
+        <define name="NAV_CIRCLE_DIST"      value="60."/>
+        <define name="NAV_LINE_DIST"        value="100"/>
+        <define name="LINE_GAIN"            value="0.2"/>
+        <define name="CLIMB_SPEED_FWD"      value="2.0"/>
+        <define name="DESCEND_SPEED_FWD"    value="-3.0"/>
+        <define name="QUADPLANE"            value="TRUE"/>
```
Comment (dewagter) [inline conf/airframes/tudelft/rot_wing_25kg.xml:450]:
```suggestion
```

File: sw/airborne/modules/rot_wing_drone/rot_wing_automation.c [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+  }
+}
+
+// Update a waypoint such that you can see on the GCS where the drone wants to go
+void update_waypoint_rot_wing_automation(uint8_t wp_id, struct FloatVect3 * target_ned) {
+
+  // Update the waypoint
+  struct EnuCoor_f target_enu;
+  ENU_OF_TO_NED(target_enu, *target_ned);
+  waypoint_set_enu(wp_id, &target_enu);
```
Comment (gautierhattenberger) [inline sw/airborne/modules/rot_wing_drone/rot_wing_automation.c:78]: It is required to add `#include "modules/nav/waypoints.h"` to pass the compilation tests.

File: conf/modules/rot_wing_automation.xml [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+    </dl_settings>
+  </settings>
+  <header>
+    <file name="rot_wing_automation.h"/>
+  </header>
+    <init fun="init_rot_wing_automation()"/>
+    <periodic fun="periodic_rot_wing_automation()" freq="10."/>
+  <makefile>
+    <file name="rot_wing_automation.c"/>
+    <test/>
```
Comment (gautierhattenberger) [inline conf/modules/rot_wing_automation.xml:23]: It is probably needed to add some defines like:
```
      <define name="DOWNLINK_TRANSPORT" value="pprz_tp"/>
      <define name="DOWNLINK_DEVICE" value="uart0"/>
      <define name="USE_UART0"/>
```

**Review by gautierhattenberger — CHANGES_REQUESTED** [review]

File: conf/airframes/tudelft/rot_wing_25kg.xml (hunk 3 above)
Comment (dewagter) [inline conf/airframes/tudelft/rot_wing_25kg.xml:121]:
```suggestion
```

File: sw/airborne/modules/rot_wing_drone/rot_wing_automation.c [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+#include "state.h"
+#include "modules/datalink/telemetry.h"
+#include "filters/low_pass_filter.h"
+#include "state.h"
+
+#include "math/pprz_algebra_float.h"
+#include "modules/nav/waypoints.h"
+
+#ifndef ROT_WING_AUTOMATION_TRANS_ACCEL
+#define ROT_WING_AUTOMATION_TRANS_ACCEL 1.0
```
Comment (fvantienen) [inline sw/airborne/modules/rot_wing_drone/rot_wing_automation.c:36]: All these defines are not documented both there and in the module file

File: conf/airframes/tudelft/rot_wing_25kg.xml [hunk 8]
```diff
@@ -393,6 +421,11 @@
     </section>
 
     <section name="GUIDANCE_INDI_HYBRID" prefix="GUIDANCE_INDI_">
+        <!--WLS settings-->
+        <define name="USE_WLS"              value="FALSE|TRUE"/>
```
Comment (dewagter) [inline conf/airframes/tudelft/rot_wing_25kg.xml:425]:
```suggestion
        <define name="USE_WLS"              value="TRUE"/>
```

File: conf/modules/rot_wing_automation.xml [hunk 9]
```diff
@@ -0,0 +1,24 @@
+<!DOCTYPE module SYSTEM "module.dtd">
+<module name="rot_wing_automation" dir="rot_wing_drone">
+  <doc>
+    <description>Fucntions to automate the navigation and guidance of the rotating wing drone</description>
```
Comment (fvantienen) [inline conf/modules/rot_wing_automation.xml:4]: Typo Functions

File: conf/modules/rot_wing_automation.xml (hunk 6 above)
Comment (dewagter) [reply conf/modules/rot_wing_automation.xml]: needs many modules to compile so removed test

File: sw/airborne/modules/rot_wing_drone/rot_wing_automation.c [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+// Update a waypoint such that you can see on the GCS where the drone wants to go
+void update_waypoint_rot_wing_automation(uint8_t wp_id, struct FloatVect3 * target_ned) {
+
+  // Update the waypoint
+  struct EnuCoor_f target_enu;
+  ENU_OF_TO_NED(target_enu, *target_ned);
+  waypoint_set_enu(wp_id, &target_enu);
+
+  // Send waypoint update every half second
+  RunOnceEvery(100/2, {
```
Comment (fvantienen) [inline sw/airborne/modules/rot_wing_drone/rot_wing_automation.c:110]: This frequency is incorrect especially if you call this function multiple times in the `rot_wing_vis_transition`.

File: sw/airborne/modules/rot_wing_drone/rot_wing_automation.c [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+  float cpsi = cosf(psi);
+  float spsi = sinf(psi);
+
+  float airspeed = stateGetAirspeed_f();
+  struct NedCoor_f *groundspeed = stateGetSpeedNed_f();
+  struct FloatVect2 airspeed_v = { cpsi * airspeed, spsi * airspeed };
+  VECT2_DIFF(rot_wing_a.windvect, *groundspeed, airspeed_v);
+
+  // Filter the wind
+  update_butterworth_2_low_pass(&rot_wing_automation_wind_filter[0], rot_wing_a.windvect.x);
```
Comment (fvantienen) [inline sw/airborne/modules/rot_wing_drone/rot_wing_automation.c:132]: `rot_wing_a.windvect_f.x =  update_butterworth_2_low_pass(&rot_wing_automation_wind_filter[0], rot_wing_a.windvect.x);`
Is nicer and works also

File: sw/airborne/modules/rot_wing_drone/rot_wing_automation.c [hunk 12]
```diff
@@ -0,0 +1,200 @@
...
+  RunOnceEvery(100/2, {
+    // Send to the GCS that the waypoint has been moved
+    DOWNLINK_SEND_WP_MOVED_ENU(DefaultChannel, DefaultDevice, &wp_id,
+                                &waypoints[wp_id].enu_i.x,
+                                &waypoints[wp_id].enu_i.y,
+                                &waypoints[wp_id].enu_i.z);
+  } );
+}
+
+void update_wind_vector(void)
```
Comment (fvantienen) [inline sw/airborne/modules/rot_wing_drone/rot_wing_automation.c:119]: Would be better to do a centralized wind measurement and broadcast this with ABI or in the state. We could move this in the future to something like `air_data`. For now this is fine.

File: sw/airborne/modules/rot_wing_drone/rot_wing_automation.c [hunk 13]
```diff
@@ -0,0 +1,200 @@
...
+
+  rot_wing_a.transitioned = false;
+  rot_wing_a.windvect.x = 0.0;
+  rot_wing_a.windvect.y = 0.0;
+  rot_wing_a.windvect_f.x = 0.0;
+  rot_wing_a.windvect_f.y = 0.0;
+
+  // Init windvector low pass filter
+  float tau = 1.0 / (2.0 * M_PI * ROT_WING_AUTOMATION_WIND_FILT_CUTOFF);
+  float sample_time = 1.0 / 10.0;
```
Comment (fvantienen) [inline sw/airborne/modules/rot_wing_drone/rot_wing_automation.c:83]: Frequency would be better to get out of the generated define

**Review by fvantienen — DISMISSED** [review]

File: sw/airborne/modules/rot_wing_drone/rot_wing_automation.c (hunk 11 above)
Comment (dewagter) [reply sw/airborne/modules/rot_wing_drone/rot_wing_automation.c:132]: good idea!

File: sw/airborne/modules/rot_wing_drone/rot_wing_automation.c (hunk 7 above)
Comment (dewagter) [reply sw/airborne/modules/rot_wing_drone/rot_wing_automation.c:37]: done

File: sw/airborne/modules/rot_wing_drone/rot_wing_automation.c (hunk 12 above)
Comment (dewagter) [reply sw/airborne/modules/rot_wing_drone/rot_wing_automation.c:123]: added to issues

File: sw/airborne/modules/rot_wing_drone/rot_wing_automation.c (hunk 13 above)
Comment (dewagter) [reply sw/airborne/modules/rot_wing_drone/rot_wing_automation.c]: done

File: sw/airborne/modules/rot_wing_drone/rot_wing_automation.c (hunk 10 above)
Comment (dewagter) [reply sw/airborne/modules/rot_wing_drone/rot_wing_automation.c:112]: This only affects the downlink.

File: sw/airborne/modules/rot_wing_drone/rot_wing_automation.c [hunk 14]
```diff
@@ -0,0 +1,198 @@
...
+
+// Update a waypoint such that you can see on the GCS where the drone wants to go
+void update_waypoint_rot_wing_automation(uint8_t wp_id, struct FloatVect3 * target_ned) {
+
+  // Update the waypoint
+  struct EnuCoor_f target_enu;
+  ENU_OF_TO_NED(target_enu, *target_ned);
+  waypoint_set_enu(wp_id, &target_enu);
+
+  // Send waypoint update every half second
```
Comment (dewagter) [inline sw/airborne/modules/rot_wing_drone/rot_wing_automation.c:111]:
```suggestion
  // Send waypoint update roughly every half second
```

**Comment by dewagter** [comment]
Fixed all comments.

**Pull request merged**